palette/moreland/smooth.go

```diff
@@ -0,0 +1,279 @@
+// December 2009. DOI 10.1007/978-3-642-10520-3_9.
+package moreland
+
+import (
+	"fmt"
+	"image/color"
+	"math"
+
+	"github.com/gonum/plot/palette"
+)
+
+// smoothDiverging is a smooth diverging color palette as described in
+// "Diverging Color Maps for Scientific Visualization." by Kenneth Moreland,
+// in Proceedings of the 5th International Symposium on Visual Computing,
+// December 2009. DOI 10.1007/978-3-642-10520-3_9.
+type smoothDiverging struct {
+	// start and end are the beginning and ending colors
+	start, end msh
+
+	// convergeM is the MSH magnitude of the convergence point.
+	// It is 88 by default.
+	convergeM float64
+
+	// alpha represents the opacity of the returned
+	// colors in the range (0,1). It is 1 by default.
+	alpha float64
+
+	// Min and Max are the minimum and maximum values of the range of
+	// scalars that can be mapped to colors using this palette.
+	min, max float64
+
+	// convergePoint is a number between min and max where the colors
+	// should converge.
+	convergePoint float64
+}
+
+// NewSmoothDiverging creates a new smooth diverging ColorMap as described in
+// "Diverging Color Maps for Scientific Visualization." by Kenneth Moreland,
+// in Proceedings of the 5th International Symposium on Visual Computing,
+// December 2009. DOI 10.1007/978-3-642-10520-3_9.
+//
+// start and end are the start- and end-point colors and
+// convergeM is the magnitude of the convergence point in
+// magnitude-saturation-hue (MSH) color space. Note that
+// convergeM specifies the color of the convergence point; it does not
+// specify the location of the convergence point.
+func NewSmoothDiverging(start, end color.Color, convergeM float64) palette.DivergingColorMap {
+	return newSmoothDiverging(colorToMSH(start), colorToMSH(end), convergeM)
+}
+
+// newSmoothDiverging creates a new smooth diverging ColorMap
+// where start and end are the start and end point colors in MSH space and
+// convergeM is the MSH magnitude of the convergence point. Note that
+// convergeM specifies the color of the convergence point; it does not
+// specify the location of the convergence point.
+func newSmoothDiverging(start, end msh, convergeM float64) palette.DivergingColorMap {
+	return &smoothDiverging{
+		start:         start,
+		end:           end,
+		convergeM:     convergeM,
+		convergePoint: math.NaN(),
+		alpha:         1,
+	}
+}
+
+// At implements the palette.ColorMap interface.
+func (p *smoothDiverging) At(scalar float64) (color.Color, error) {
+	if p.min == p.max {
+		return nil, fmt.Errorf("moreland: DivergingMSH color map Max == Min")
+	}
+	if p.min > p.max {
+		return nil, fmt.Errorf("moreland: DivergingMSH color map Max < Min")
+	}
+	convergePoint := (p.convergePoint - p.min) / (p.max - p.min)
+	scalar = (scalar - p.min) / (p.max - p.min)
+	o := p.interpolateMSHDiverging(scalar, convergePoint).cieLAB().cieXYZ().rgb().sRGBA(p.alpha)
+	if o.R > 1 || o.G > 1 || o.B > 1 || o.A > 1 || o.R < 0 || o.G < 0 || o.B < 0 || o.A < 0 {
```

Comment:
There previously was one as a method for sRGBA, but I was asked incorporate it into this function because it is only used in one place